contributors/design-proposals/scheduling/pod-priority-resourcequota.md

```diff
@@ -0,0 +1,186 @@
+        // If the object in the intercepted request matches these rules,
+        // quota system will ensure that corresponding quota MUST have
+        // priority based Scopes matching the object in request. 
+        //
+        // If MatchScopes has matched on an object, request for the resource will be denied 
+        // if there is no quota with matching Scopes. In this case, matching priority class based Scopes
+        // will be an additional requirement for any quota to qualified as covering quota.
+        // +optional
+        MatchScopes []string `json:"matchScopes,omitempty"`
+}
+```
+
+### Expected Behavior of ResourceQuota Admission Controller and ResourceQuota System
+`MatchScopes` will be configured in admission controller configuration to apply quota based on priority class names. If `MatchScopes` matches/selects an incoming pod request, request will be **denied if a Covering Quota is missing**. The meaning of Covering Quota is: any quota which has priority class based `Scopes` that matches/selects the pod in the request. 
+
+Please note that this priority class based criteria will be an **additional** criteria that must be satisfied by covering quota.
+
+For more details, please refer to the `Sample user story` sections at the end of this doc.
+
+#### Backward Compatibility
+
+If a Pod's requested resources are not matched by any of the filters in admission controller configuration's `MatchScopes`, overall behavior for the pod will be same as it is today where `ResourceQuota` has no awareness of priority. In such a case, request will be allowed if no covering `ResourceQuota` is found.
+
+Couple of other noteworthy details:
+1. If multiple `ResourceQuota` apply to a Pod, the pod must satisfy all of them.
+2. We do not enforce referential integrity across objects. i.e. Creation or updating of ResourceQuota object, scopes of which names a PriorityClass that does not exist, are allowed.
+
+This design also tries to enable flexibility for its configuration. Here are several sample user stories.
+
+#### Sample User Story 1
+**As a cluster admin, I want `cluster-services` priority only apply to `kube-system` namespace , so that I can ensure those critical daemons on each node while normal user's workloads will not disrupt that ability.**
+
+To enforce above policy:
+1. Admin will create admission controller configuration as below:
+```yaml
+apiVersion: apiserver.k8s.io/v1alpha1
+kind: AdmissionConfiguration
+plugins:
+- name: "ResourceQuota"
+  configuration:
+  apiVersion: resourcequota.admission.k8s.io/v1alpha1
+  kind: Configuration
+  limitedResources:
+  - resource: pods
+    matchScopes:
+    - "ResourceQuotaScopePriorityClassNameIn:cluster-services"
+```
+
+2. Admin will then create a corresponding resource quota object in `kube-system` namespace:
+
+    `$ kubectl create quota critical --hard=count/pods=10 --scopes=ResourceQuotaScopePriorityClassNameIn:cluster-services -n kube-system`
```

Comment:
As per slack conversation, the question is 
- What if `cluster-services` is replaced with one of `system-cluster-critical` or `system-node-critical`, these won't be available resolvable at API level as they are coming from https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/admission/priority/admission.go#L49.

2 approaches: 
- (Less elegant)Move the `SystemPriorityClasses` map up till pkg/apis and refer it from there. If priorityClass name is not found while listing, check if the entry exists in map and if not found deny pod creation. The same map could be used in admission controller.
- (Recommended - suggested by @derekwaynecarr @liggitt) auto-create `system-cluster-critical` and `system-node-critical` at startup time just like other priorityClasses if priority is enabled. This ensures that listing of priorityClasses also returns system-cluster-critical and system-node-critical. I believe @bsalamat is also ok with this approach.

/cc @aveshagarwal

Reply:
i just want to ensure that we have no magic priority class names, and that any priority class names we reserve are manifested in the API server storage.  its tangential to this proposal but from an admin experience, not seeing priorities returned in a kubectl command will make it hard for them to know how to configure admission.